daemon/prompter.c

```diff
@@ -996,9 +1050,124 @@ prompter_cancel_invocation(prompter_t *self)
+prompter_watch_name(prompter_t *self, GDBusConnection *connection, const gchar *name)
+{
+    if( !g_hash_table_contains(self->prm_watchers, name) ) {
+        guint id = g_bus_watch_name_on_connection(
+                connection, name, G_BUS_NAME_OWNER_FLAGS_NONE, NULL,
+                prompter_handle_name_lost_cb, self, NULL);
+        g_hash_table_insert(self->prm_watchers, g_strdup(name), GUINT_TO_POINTER(id));
+        prompter_name_has_owner(self, connection, name);
+        log_debug("watching for '%s' to leave bus", name);
     }
 }
 
+static void
+prompter_unwatch_name(prompter_t *self, const gchar *name)
+{
+    guint watcher = GPOINTER_TO_UINT(g_hash_table_lookup(self->prm_watchers, name));
+    if( watcher )
+        g_bus_unwatch_name(watcher);
+    g_hash_table_remove(self->prm_watchers, name);
+}
+
+static void
+prompter_handle_name_lost(prompter_t *self, const gchar *name)
+{
+    prompter_unwatch_name(self, name);
+
+    /* First check the current invocation */
+    GDBusMethodInvocation *invocation = self->prm_invocation;
+    if( invocation && !g_strcmp0(g_dbus_method_invocation_get_sender(invocation), name) ) {
+        log_debug("-> canceling %p", invocation);
+        prompter_return_error(invocation, G_DBUS_ERROR_AUTH_FAILED,
+                              SERVICE_MESSAGE_DISCONNECTED);
+        self->prm_invocation = NULL;
+        prompter_set_state(self, PROMPTER_STATE_CANCEL);
+    }
+
+    /* Then check the rest of the invocations */
+    GList *iter = prompter_iter(self);
+    while( iter ) {
+        invocation = iter->data;
+        if( !g_strcmp0(g_dbus_method_invocation_get_sender(invocation), name) ) {
+            log_debug("-> skipping %p", invocation);
+            prompter_return_error(invocation, G_DBUS_ERROR_AUTH_FAILED,
+                                  SERVICE_MESSAGE_DISCONNECTED);
+            iter = prompter_drop(self, iter);
+        }
+        else {
+            iter = iter->next;
+        }
+    }
+}
+
+static void
+prompter_handle_name_lost_cb(GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+    (void)connection; // unused
+    prompter_t *self = user_data;
+    log_debug("'%s' left bus", name);
+    prompter_handle_name_lost(self, name);
+}
+
+static void
+prompter_name_has_owner(prompter_t *self, GDBusConnection *connection, const gchar *name)
+{
+    gpointer *data = g_malloc_n(2, sizeof *data);
+    data[0] = self;
+    data[1] = g_strdup(name);
+    g_dbus_connection_call(connection,
+                           DBUS_SERVICE,
+                           DBUS_PATH,
+                           DBUS_INTERFACE,
+                           DBUS_METHOD_NAME_HAS_OWNER,
+                           g_variant_new("(s)", name),
+                           NULL,
+                           G_DBUS_CALL_FLAGS_NONE,
+                           -1,
+                           NULL,
+                           prompter_name_has_owner_cb,
+                           data);
```

Comment:
Well, let's suppose there isn't any. One option would be to make this proper with constructors and everything and store this data in the GHashTable together with the watcher id and a pointer back to prompter. Something like `struct name_watcher_t`. How does that sound to you?

Reply:
Added `watcher_t`. I think it made this somewhat cleaner but please check. There is one hazardous place where callback gets called with pointer to freed memory but it's difficult to avoid and we can anyway see that from the error. There is a comment there that the pointer may be invalid.

Reply:
Overall, yes, much clearer.